embulk-core/src/main/java/org/embulk/exec/BulkLoader.java

```diff
@@ -322,28 +322,34 @@ public RuntimeException getRepresentativeException()
 
         public ExecutionResult buildExecuteResult()
         {
-            return buildExecuteResultWithWarningException(null);
+            return buildExecuteResultWithWarningException(false, null);
         }
 
-        public ExecutionResult buildExecuteResultWithWarningException(Throwable ex)
+        public ExecutionResult buildExecuteResultWithWarningException(boolean skipped, Throwable ex)
```

Comment:
How about adding another method rather than making this method type-unsafe?

```
diff --git a/embulk-core/src/main/java/org/embulk/exec/BulkLoader.java b/embulk-core/src/main/java/org/embulk/exec/BulkLoader.java
index 9de2a20..c7d00d2 100644
--- a/embulk-core/src/main/java/org/embulk/exec/BulkLoader.java
+++ b/embulk-core/src/main/java/org/embulk/exec/BulkLoader.java
@@ -322,34 +322,38 @@ public class BulkLoader
 
         public ExecutionResult buildExecuteResult()
         {
-            return buildExecuteResultWithWarningException(false, null);
+            return buildExecuteResultWithWarningException(null);
         }
 
-        public ExecutionResult buildExecuteResultWithWarningException(boolean skipped, Throwable ex)
+        public ExecutionResult buildExecuteResultWithWarningException(Throwable ex)
         {
-            ConfigDiff configDiff;
-            if (skipped) {
-                configDiff = ((SkipTransactionException) ex).getConfigDiff();
+            ConfigDiff configDiff = Exec.newConfigDiff();
+            if (inputConfigDiff != null) {
+                configDiff.getNestedOrSetEmpty("in").merge(inputConfigDiff);
             }
-            else {
-                configDiff = Exec.newConfigDiff();
-                if (inputConfigDiff != null) {
-                    configDiff.getNestedOrSetEmpty("in").merge(inputConfigDiff);
-                }
-                if (outputConfigDiff != null) {
-                    configDiff.getNestedOrSetEmpty("out").merge(outputConfigDiff);
-                }
+            if (outputConfigDiff != null) {
+                configDiff.getNestedOrSetEmpty("out").merge(outputConfigDiff);
             }
 
             ImmutableList.Builder<Throwable> ignoredExceptions = ImmutableList.builder();
             for (Throwable e : getExceptions()) {
                 ignoredExceptions.add(e);
             }
-            if (ex != null && !skipped) {
+            if (ex != null) {
                 ignoredExceptions.add(ex);
             }
 
-            return new ExecutionResult(configDiff, skipped, ignoredExceptions.build());
+            return new ExecutionResult(configDiff, false, ignoredExceptions.build());
+        }
+
+        public ExecutionResult buildExecuteResultOfSkippedExecution(ConfigDiff configDiff)
+        {
+            ImmutableList.Builder<Throwable> ignoredExceptions = ImmutableList.builder();
+            for (Throwable e : getExceptions()) {
+                ignoredExceptions.add(e);
+            }
+
+            return new ExecutionResult(configDiff, true, ignoredExceptions.build());
         }
 
         public ResumeState buildResumeState(ExecSession exec)
@@ -571,10 +575,12 @@ public class BulkLoader
             return state.buildExecuteResult();
 
         } catch (Throwable ex) {
-            boolean skipped = isSkippedTransaction(ex);
-            if (skipped || (state.isAllTasksCommitted() && state.isAllTransactionsCommitted())) {
+            if (isSkippedTransaction(ex)) {
+                ConfigDiff configDiff = ((SkipTransactionException) ex).getConfigDiff();
+                return state.buildExecuteResultOfSkippedExecution(configDiff);
+            } else if (state.isAllTasksCommitted() && state.isAllTransactionsCommitted()) {
                 // ignore the exception
-                return state.buildExecuteResultWithWarningException(skipped, ex);
+                return state.buildExecuteResultWithWarningException(ex);
             }
             throw state.buildPartialExecuteException(ex, Exec.session());
         }
@@ -649,10 +655,12 @@ public class BulkLoader
             return state.buildExecuteResult();
 
         } catch (Throwable ex) {
-            boolean skipped = isSkippedTransaction(ex);
-            if (skipped || (state.isAllTasksCommitted() && state.isAllTransactionsCommitted())) {
+            if (isSkippedTransaction(ex)) {
+                ConfigDiff configDiff = ((SkipTransactionException) ex).getConfigDiff();
+                return state.buildExecuteResultOfSkippedExecution(configDiff);
+            } else if (state.isAllTasksCommitted() && state.isAllTransactionsCommitted()) {
                 // ignore the exception
-                return state.buildExecuteResultWithWarningException(skipped, ex);
+                return state.buildExecuteResultWithWarningException(ex);
             }
             throw state.buildPartialExecuteException(ex, Exec.session());
         }
```